bftengine/src/bftengine/ControlStateManager.cpp

```diff
@@ -14,12 +14,34 @@
 #include "ControlStateManager.hpp"
 namespace bftEngine {
 
-void ControlStateManager::setStopAtNextCheckpoint() {}
+// This method will be called from the command handler. Notice that the command handler does not aware to the number of
+// the checkpoint. Thus, it can only raise a flag to stop.
+void ControlStateManager::setStopAtNextCheckpoint(int64_t currentSeqNum) {
+  uint64_t seq_num_to_stop_at = (currentSeqNum + 2 * checkpointWindowSize);
+  seq_num_to_stop_at = seq_num_to_stop_at - (seq_num_to_stop_at % checkpointWindowSize);
+  std::ostringstream outStream;
+  controlStateMessages::StopAtNextCheckpointMessage msg(seq_num_to_stop_at);
+  concord::serialize::Serializable::serialize(outStream, msg);
+  auto data = outStream.str();
+  state_transfer_->saveReservedPage(
+      resPageOffset() + reserved_pages_indexer_.update_reserved_page_, data.size(), data.data());
```

Comment:
Can we please encapsulate `resPageOffset() + reserved_pages_indexer_.update_reserved_page_` in a method and explain what it's doing with a header comment if necessary.

Reply:
done